hdfs/client.py

```diff
@@ -76,22 +77,45 @@ def to_method(self, operation):
 
     def api_handler(client, hdfs_path, data=None, strict=True, **params):
       """Wrapper function."""
-      url = '%s%s%s' % (
-        client.url.rstrip('/'),
-        self.webhdfs_prefix,
-        quote(client.resolve(hdfs_path), '/= '),
-      )
       params['op'] = operation
       if client._proxy is not None:
         params['doas'] = client._proxy
-      return client._request(
-        method=self.method,
-        url=url,
-        data=data,
-        params=params,
-        strict=strict,
-        **self.kwargs
-      )
+
+      # Try each URL configured in the client
+      # If we get a socket error, rotate the URLs
+      # If we get a HDFS error, check for standby and rotate if so
+      # If don't get a successful response, raise the last occurring exception
+      max_attempts = len(client._urls)
+      exc = None
+      for _ in range(max_attempts):
+        url = '%s%s%s' % (
+          client._urls[0].rstrip('/'),
+          self.webhdfs_prefix,
+          quote(client.resolve(hdfs_path), '/= '),
+        )
+        try:
+          return client._request(
+            self.method,
+            url=url,
+            data=data,
+            params=params,
+            strict=strict,
+            **self.kwargs
+          )
+        except (rq.exceptions.ReadTimeout,
+                  rq.exceptions.ConnectTimeout,
+                  rq.exceptions.ConnectionError,
+                  ) as e:
+          exc = e
+          client._urls.rotate(-1)
```

Comment:
This won't quite work when multiple threads are involved. The client class was originally thread-safe so we should try to maintain it.

Here's a rough suggestion:

+ First, we add a `threading.Lock()` on each client instance, say `client._lock`.
+ Then, we re-introduce the `HdfsStandbyError` (it makes exception handling a bit simpler).
+ Finally, we update the code here as follows:

```python
attempted_hosts = set()
hosts = client._urls
while True:
  with client._lock:
    while hosts[0] in attempted_hosts:
      hosts.rotate(-1)
    host = hosts[0]
  url = '%s%s%s' % (
    host.rstrip('/'),
    self.webhdfs_prefix,
    quote(client.resolve(hdfs_path), '/= '),
  )
  try:
    return client._request(
      method=self.method,
      url=url,
      data=data,
      params=params,
      strict=strict,
      **self.kwargs
    )
  except (ReadTimeout, ConnectTimeout, ConnectionError, HdfsStandbyError) as exc:
    attempted_hosts.add(host)
    if len(attempted_hosts) == len(hosts):
      raise exc
```

Let me know what you think.